File: lc2563.java
import java.util.Arrays;

class LC2563 {
    static int lessEqual(int index, int target, int[] arr) {
        int l = index;
        int p = -1;
        int r = arr.length - 1;
        int mid;
        while (l <= r) {
            mid = l + (r - l) / 2;
            if (arr[mid] > target) {
                r = mid - 1;
            } else {
                p = mid;
                l = mid + 1;
            }
        }
        if (p == -1) return -1;
        
        return p;
    }

    static int lessThan(int index, int target, int[] arr) {
        int l = index;
        int p = -1;
        int r = arr.length - 1;
        int mid;
        while (l <= r) {
            mid = l + (r - l) / 2;
            if (arr[mid] >= target) {
                r = mid - 1;
            } else {
                p = mid;
                l = mid + 1;
            }
        }
        if (p == -1) return -1;
        
        return p;
    }

    public long countFairPairs(int[] nums, int lower, int upper) {
        // count the pairs lower than LOWER
        // count the pairs lower and equal to UPPER
        // subtract them
        Arrays.sort(nums);
        long sum=0;
        for (int i = 0; i < nums.length; ++i) {
            int tmp1= lessThan(i + 1,lower - nums[i], nums);
            int tmp2= lessEqual(i + 1,upper - nums[i], nums);
            // problems for  {-5,-7,-5,-7,-5} input.
            int pairsSmallerThanLower= tmp1==-1?0:tmp1-i;
            int pairsSmallerEqualToUpper=tmp2==-1?0: tmp2-i;
            if(pairsSmallerEqualToUpper==0 & pairsSmallerThanLower==0) break;
            sum+=pairsSmallerEqualToUpper-pairsSmallerThanLower;
        }
        return sum;

    }

    public static void main(String[] args) {
        LC2563 s = new LC2563();
        // int [] arr= {0,1,2,3,7,4,4,5};
        int[] arr = { 0, 1, 7, 4, 4, 5 };
        int[] arr1 = {-5,-7,-5,-7,-5};
        // s.countFairPairs(arr, 3, 6);
        s.countFairPairs(arr1, -12, -12);
        Arrays.sort(arr1);
        
        for (int i : arr1) {
            System.out.print(i+" ");
            
        }
        System.out.println();
        System.out.println(lessEqual(0, -7, arr1));
        System.out.println(lessThan(0, -3, arr1));
    }

}

File: lc2563.py
# can be sorted order doesn't matter
#[0, 1, 4, 4, 5, 7]
#lower =3 ,upper =6
# for 0 -> 4,4,5,
#p= for 0 how many lower than 3?-> 1
# q=for 0 how many lower than equal 6?-> 1,4,4,5
# so between = q-p


from typing import List


class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        #returns the first number that is exactly lower than the provided
        # find me the index 
        N = len(nums)
        cnt =0
        nums= sorted(nums)
        def lowerBound(start, end , target):
            idx=-1
            left =start
            right =end
            while left <N and right<N and left<= right:
                mid = (left+right)//2
                if(nums[mid] >= target):
                    right= mid -1
                else:
                    idx = mid
                    left = mid +1
            return idx
        # find the first num thats bigger than target
        def upperBound(start, end , target):
            idx=-1
            left =start
            right =end
            while left <N and right<N and left<= right:
                mid = (left+right)//2
                if(nums[mid] <= target):
                    left= mid + 1
                else:
                    idx = mid
                    right = mid -1

            return idx

        for i in range(N):
            # lower than lower
            tmp1=lowerBound(i+1,N-1,lower-nums[i] )
            if tmp1 ==-1 :
                p=0
            else: p = tmp1 -i

            tmp2=upperBound(i+1,N-1, upper-nums[i] ) 
            if tmp2==-1:
                q = N -i -1
            else : 
                q = tmp2 -i -1

            if q-p<0:
                continue
            cnt+= q-p


        return cnt
        
s=Solution()
s.countFairPairs(nums = [0,1,7,4,4,5], lower = 3, upper = 6)
s.countFairPairs(nums = [1,7,9,2,5], lower = 11, upper = 11)
